File: raytracer_rust_common/src/components/texture.rs
use std::sync::Arc;

use enum_dispatch::enum_dispatch;

use crate::{
    components::{
        noise::{Noise, NoiseEnum},
        surface::SurfaceHit,
    },
    util::vec3::{Color, color},
};

#[enum_dispatch]
#[derive(Debug, Clone)]
pub enum TextureEnum {
    Shared,
    SolidColor,
    Checker,
    NoiseTexture,
}

#[enum_dispatch(TextureEnum)]
pub trait Texture: Into<TextureEnum> {
    fn value(&self, hit: &SurfaceHit<'_>) -> Color;

    fn shared(self) -> Shared
    where
        Self: std::marker::Sized,
    {
        Shared {
            inner: Arc::new(self.into()),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Shared {
    inner: Arc<TextureEnum>,
}

impl Texture for Shared {
    fn value(&self, hit: &SurfaceHit<'_>) -> Color {
        self.inner.value(hit)
    }
}

#[derive(Debug, Clone)]
pub struct SolidColor {
    albedo: Color,
}

impl SolidColor {
    pub fn new(albedo: Color) -> Self {
        Self { albedo }
    }
}

impl Texture for SolidColor {
    fn value(&self, _hit: &SurfaceHit<'_>) -> Color {
        self.albedo
    }
}

impl From<Color> for TextureEnum {
    fn from(albedo: Color) -> Self {
        SolidColor::new(albedo).into()
    }
}

#[derive(Debug, Clone)]
pub struct Checker {
    inverse_scale: f64,
    even: Box<TextureEnum>,
    odd: Box<TextureEnum>,
}

impl Checker {
    pub fn new(scale: f64, even: impl Into<TextureEnum>, odd: impl Into<TextureEnum>) -> Self {
        Self {
            inverse_scale: 1.0 / scale,
            even: Box::new(even.into()),
            odd: Box::new(odd.into()),
        }
    }
}

impl Texture for Checker {
    fn value(&self, hit: &SurfaceHit<'_>) -> Color {
        let p = (self.inverse_scale * hit.point).floor();
        let is_even = (p.x() as i32 + p.y() as i32 + p.z() as i32) % 2 == 0;

        if is_even {
            self.even.value(hit)
        } else {
            self.odd.value(hit)
        }
    }
}

#[derive(Debug, Clone)]
pub struct NoiseTexture {
    noise: NoiseEnum,
    scale: f64,
}

impl NoiseTexture {
    pub fn new(noise: impl Into<NoiseEnum>, scale: f64) -> Self {
        Self {
            noise: noise.into(),
            scale,
        }
    }
}

impl Texture for NoiseTexture {
    fn value(&self, hit: &SurfaceHit<'_>) -> Color {
        color(1.0, 1.0, 1.0) * 0.5 * (1.0 + self.noise.noise(self.scale * hit.point))
    }
}

File: raytracer_ts/texture.ts
import type { Noise } from './noise';
import { Vec3, type Color3, type Point3 } from './util';

export abstract class Texture {
    public abstract value(u: number, v: number, point: Point3): Color3;
}

export class SolidColor extends Texture {
    constructor(private albedo: Color3) {
        super();
    }

    public value(u: number, v: number, point: Point3): Color3 {
        return this.albedo;
    }
}

export class Checker extends Texture {
    private invertedScale: number;
    constructor(
        scale: number,
        private even: Texture,
        private odd: Texture,
    ) {
        super();
        this.invertedScale = 1 / scale;
    }

    public value(u: number, v: number, point: Point3): Color3 {
        return point //
            .map((n) => Math.floor(this.invertedScale * n))
            .fold((x, y, z) => (x + y + z) % 2 === 0)
            ? this.even.value(u, v, point)
            : this.odd.value(u, v, point);
    }
}

export class NoiseTexture extends Texture {
    constructor(
        private noise: Noise,
        private scale: number,
    ) {
        super();
    }

    public value(u: number, v: number, point: Point3): Color3 {
        const noise = this.noise.noise(point.times(this.scale));
        return Vec3.one.times(0.5 * (1 + noise));
    }
}

export class TurbulenceTexture extends Texture {
    constructor(
        private noise: Noise,
        private scale: number,
    ) {
        super();
    }

    public value(u: number, v: number, point: Point3): Color3 {
        return Vec3.of(0.5).times(1 + Math.sin(this.scale * point.z + 10 * this.noise.turbulence(point, 7)));
    }
}
